src/AutoRest.CSharp/Common/Output/Builders/JsonSerializationMethodsBuilder.cs

```diff
@@ -195,6 +210,64 @@ private static MethodBodyStatement[] WriteObject(JsonObjectSerialization seriali
                 utf8JsonWriter.WriteEndObject()
             };
 
+        private static MethodBodyStatement[] BuildJsonModelWriteMethodBody(Method jsonModelWriteCoreMethod, Utf8JsonWriterExpression utf8JsonWriter)
+        {
+            var coreMethodSignature = jsonModelWriteCoreMethod.Signature;
+
+            return new[]
+            {
+                utf8JsonWriter.WriteStartObject(),
+                This.Invoke(coreMethodSignature.Name, coreMethodSignature.Parameters.Select(p => (ValueExpression)p).ToList()).ToStatement(),
```

Comment:
we should be able just use
```csharp
This.Invoke(coreMethodSignature)
```
there is an overload taking a signature

Reply:
The overload is taking `MethodSignature`, not `MethodSignatureBase`.

Reply:
then cast it to MethodSignature.

Reply:
What benefit we could get from cast? I don't want to do casting. It makes me feel unsafe.

Reply:
In MGC, the signature from a "method" is `MethodSignature`.
This is a bad design on our side that the constructor method and normal method share the same type `Method` which requires that we should always "instantiate a Method with a MethodSignature, instantiate a constructor with a ConstructorSignature", therefore it is very safe we do the cast here because we know it should never be a constructor here.
In MGC, we changed it so that constructor has its own type `ConstructorProvider` and normal method has its own independent type `MethodProvider`.

Back to this comment, it would be cleaner if we use the cast here, this is the benefit. And the disadvantage you want to avoid is that you are afraid the signature we get here is not `MethodSignature`, but this will never happen, or if it happens, we should be happy that it blows up here because we need to know it goes wrong.